src/lib/comms/sol-netctl-impl-connman.c

```diff
@@ -0,0 +1,1133 @@
+        return -EINVAL;
+    }
+
+    _ctx.connman = sol_bus_client_new(bus, "org.connman");
+    if (!_ctx.connman) {
+        sd_bus_unref(bus);
+        SOL_WRN("Unable to new a bus client\n");
+        return -EINVAL;
+    }
+
+    sol_ptr_vector_init(&_ctx.service_vector);
+    sol_monitors_init(&_ctx.service_ms, NULL);
+    sol_monitors_init(&_ctx.manager_ms, NULL);
+    sol_monitors_init(&_ctx.error_ms, NULL);
+
+    return 0;
+}
+
+static void
+sol_connman_shutdown_lazy(void)
+{
+    struct sol_connman_service *service;
+    uint16_t id;
+
+    _ctx.refcount--;
+
+    if (_ctx.refcount)
+        return;
+
+    if (_ctx.connman) {
+        sol_bus_client_free(_ctx.connman);
+        _ctx.connman = NULL;
+    }
+
+    _ctx.state_slot =
+        sd_bus_slot_unref(_ctx.state_slot);
+    _ctx.manager_slot =
+        sd_bus_slot_unref(_ctx.manager_slot);
+    _ctx.service_slot =
+        sd_bus_slot_unref(_ctx.service_slot);
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&_ctx.service_vector, service, id)
+        _free_connman_service(service);
+
+    sol_ptr_vector_clear(&_ctx.service_vector);
+    sol_monitors_clear(&_ctx.service_ms);
+    sol_monitors_clear(&_ctx.manager_ms);
+    sol_monitors_clear(&_ctx.error_ms);
+
+    _ctx.connman_state = SOL_CONNMAN_STATE_UNKNOWN;
+}
+
+static int
+_match_properties_changed(sd_bus_message *m, void *userdata,
+    sd_bus_error *ret_error)
+{
+    const char *interface;
+    int r;
+
+    SOL_NULL_CHECK(_ctx.properties_changed, -EINVAL);
+
+    interface = sd_bus_message_get_interface(m);
+    SOL_NULL_CHECK(interface, -EINVAL);
+
+    if (!strstartswith(interface, "net.connman."))
+        return -EINVAL;
+
+    if (sd_bus_message_is_signal(m, "net.connman.Manager",
+        "ServicesChanged")) {
+        if (_ctx.service_slot)
+            return -EINVAL;
+
+        services_list_changed(m);
+    } else if (sd_bus_message_is_signal(m, "net.connman.Manager",
+        "PropertyChanged")) {
+        char *str = NULL;
+
+        if (_ctx.manager_slot)
+            return -EINVAL;
+
+        r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &str);
+        SOL_INT_CHECK(r, < 0, r);
+        if (streq(str, "State")) {
+            r = get_manager_properties(m);
+            SOL_INT_CHECK(r, < 0, r);
+            call_manager_monitor_callback();
+        } else {
+            SOL_DBG("Ignored changed property: %s", str);
+            r = sd_bus_message_skip(m, "v");
+            SOL_INT_CHECK(r, < 0, r);
+        }
+    }
+
+    return 0;
+}
+
+static int
+dbus_service_add_monitor(
+    sol_connman_service_monitor_cb cb, const void *data)
+{
+    int r;
+    struct sol_monitors_entry *e;
+    char matchstr[256];
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+
+    SOL_NULL_CHECK(bus, -EINVAL);
+
+    r = sol_monitors_find(&_ctx.service_ms, (sol_monitors_cb_t)cb, data);
+    if (r < 0) {
+        e = sol_monitors_append(&_ctx.service_ms, (sol_monitors_cb_t)cb, data);
```

Comment:
for consistency with other parts of soletta (ie: https://github.com/solettaproject/soletta/blob/master/src/lib/common/sol-platform.c) do not check if the cb/data exists, only append. Do this for all your monitors